### PR TITLE
Add `list_cosine_similarity`, `list_distance`, `list_inner_product`

### PR DESCRIPTION
**Pull request by Maxxen**
This PR adds two new scalar functions:
- `list_distance(FLOAT[], FLOAT[]) -> FLOAT`/`list_distance(DOUBLE[], DOUBLE[]) -> DOUBLE` 
- `list_cosine_similarity(FLOAT[], FLOAT[]) -> FLOAT`/`list_cosine_similarity(DOUBLE[], DOUBLE[]) -> DOUBLE` 

To compute the cartesian distance and cosine similarity between two lists. There are also two binary operator aliases `<=>` and `<->` for cosine similarity and distance respectively following pgvector's convention.

These functions are primarily useful for vector similarity search, but note that DuckDB's list vectors are not optimal to store embeddings in since accessing their elements require quite a bit of indirection in order for them to be able to handle different size lists and null values, dictionary compression etc. For high-performance vector search I'd advise you to look elsewhere.

Although I've tried to optimize these functions somewhat. They will throw if any list child vector contains null values in order to avoid having to check the validity of every element (also not sure what else would make sense to do?) and should also handle execution over constant/dictionary vectors without flattening.

### CODE REVIEW

**Comment by Maxxen** [comment]
Alright I couldn't help myself, I've added `list_inner_product` with a pair of accompanying `list_dot_product` and `<#>` aliases as well.

File: src/core_functions/scalar/list/list_distance.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+	auto count = args.size();
+	auto &left = args.data[0];
+	auto &right = args.data[1];
+	auto left_count = ListVector::GetListSize(left);
+	auto right_count = ListVector::GetListSize(right);
+
+	UnifiedVectorFormat left_child_format;
+	ListVector::GetEntry(left).ToUnifiedFormat(left_count, left_child_format);
+	if (!left_child_format.validity.CheckAllValid(left_count)) {
```
Comment (Mytherin) [inline src/core_functions/scalar/list/list_distance.cpp:18]: The `ToUnifiedFormat` here is somewhat misleading as the selection vector is ignored - neither in `CheckAllValid` nor in the main loop. As a result this code only works when the list child is a flat vector. I believe we enforce list children to be flat vectors now always anyway so it doesn't matter - but perhaps we can be explicit about this and use the `FlatVector` methods instead?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - one comment:

File: src/core_functions/scalar/list/list_distance.cpp (hunk 1 above)
Comment (Maxxen) [reply src/core_functions/scalar/list/list_distance.cpp]: Good point, I really just wanted some way to access the validity. I've swapped to the FlatVector accessors and added asserts for the child vector type, but AFAIK you're right about list children always being flat.

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**